hudi-client/hudi-client-common/src/test/java/org/apache/hudi/client/transaction/TestTransactionManager.java

```diff
@@ -0,0 +1,198 @@
+    writer1.start();
+
+    // Writer2 will block on trying to acquire the lock
+    // and will eventually get the lock before the timeout.
+    Thread writer2 = new Thread(() -> {
+      latch.countDown();
+      try {
+        latch.await(awaitMaxTimeoutMs, TimeUnit.MILLISECONDS);
+      } catch (InterruptedException e) {
+        //
+      }
+      assertDoesNotThrow(() -> {
+        transactionManager.beginTransaction();
+      });
+      assertDoesNotThrow(() -> {
+        transactionManager.endTransaction();
+      });
+      writer2Completed.set(true);
+    });
+    writer2.start();
+
+    // Let writer1 and writer2 wait at the sync up
+    // point to make sure they run in parallel and
+    // one get blocked by the other.
+    latch.countDown();
+    try {
+      writer1.join();
+      writer2.join();
+    } catch (InterruptedException e) {
+      //
+    }
+
+    // Make sure both writers actually completed good
+    Assertions.assertTrue(writer1Completed.get());
+    Assertions.assertTrue(writer2Completed.get());
+  }
+
+  @Test
+  public void testTransactionsWithInstantTime() {
+    // 1. Begin and end by the same transaction owner
+    Option<HoodieInstant> lastCompletedInstant = getInstant("0000001");
+    Option<HoodieInstant> newTxnOwnerInstant = getInstant("0000002");
+    transactionManager.beginTransaction(newTxnOwnerInstant, lastCompletedInstant);
+    Assertions.assertTrue(transactionManager.getCurrentTransactionOwner() == newTxnOwnerInstant);
+    Assertions.assertTrue(transactionManager.getLastCompletedTransactionOwner() == lastCompletedInstant);
+    transactionManager.endTransaction(newTxnOwnerInstant);
+    Assertions.assertFalse(transactionManager.getCurrentTransactionOwner().isPresent());
+    Assertions.assertFalse(transactionManager.getLastCompletedTransactionOwner().isPresent());
+
+    // 2. Begin transaction with a new txn owner, but end transaction with no/wrong owner
+    lastCompletedInstant = getInstant("0000002");
+    newTxnOwnerInstant = getInstant("0000003");
+    transactionManager.beginTransaction(newTxnOwnerInstant, lastCompletedInstant);
+    transactionManager.endTransaction();
+    // Owner reset would not happen as the end txn was invoked with an incorrect current txn owner
+    Assertions.assertTrue(transactionManager.getCurrentTransactionOwner() == newTxnOwnerInstant);
+    Assertions.assertTrue(transactionManager.getLastCompletedTransactionOwner() == lastCompletedInstant);
+
+    // 3. But, we should be able to begin a new transaction for a new owner
+    lastCompletedInstant = getInstant("0000003");
+    newTxnOwnerInstant = getInstant("0000004");
+    transactionManager.beginTransaction(newTxnOwnerInstant, lastCompletedInstant);
+    Assertions.assertTrue(transactionManager.getCurrentTransactionOwner() == newTxnOwnerInstant);
+    Assertions.assertTrue(transactionManager.getLastCompletedTransactionOwner() == lastCompletedInstant);
+    transactionManager.endTransaction(newTxnOwnerInstant);
```

Comment:
can you help me understand how this succeeds? bcoz, lock provider is no-op or are we using inProcessLockProvider? If a valid lock provider is used, how come  a new owner is able to begin trnx when another instant has already acquired the lock ? Or is it that, its the same thread and so it succeeds ?